direct_failure_detector/failure_detector.cc

```diff
@@ -0,0 +1,699 @@
+        fd._impl->add_listener(l, threshold, shard);
+    });
+
+    co_return s;
+}
+
+subscription::subscription(failure_detector& fd, listener& l) noexcept
+        : _fd(fd), _listener(&l) {
+}
+
+subscription::subscription(subscription&& s) noexcept
+        : _fd(s._fd), _listener(s._listener) {
+    // So the moved-from subscription does not deregister the listener when destroyed.
+    s._listener = nullptr;
+}
+
+subscription::~subscription() {
+    if (!_listener) {
+        // We were moved from, nothing to do.
+        return;
+    }
+
+    // Start by removing the listener from _registered set which prevents the failure detector from dereferencing the listener.
+    if (!_fd._impl->_registered.erase(_listener)) {
+        return;
+    }
+
+    // Cleaning up the data structures on each shard happens in the background.
+    //
+    // If we immediately register the same listener after deregistering it, we may have a problem.
+    // Hence we require each listener to be registered at most once.
+    _fd._impl->_destroy_subscriptions = _fd._impl->_destroy_subscriptions.then([l_ = _listener, &fd = _fd] () -> future<> {
+        auto l = l_;
+        try {
+            co_await fd.container().invoke_on_all([l] (failure_detector& fd) {
+                fd._impl->remove_listener(l);
+            });
+        } catch (...) {
+            logger.error("unexpected exception when deregistering listener {}: {}", fmt::ptr(l), std::current_exception());
+        }
+    });
+}
+
+void failure_detector::impl::add_listener(listener_id id, clock::interval_t threshold, seastar::shard_id shard) {
+    if (!_shard_workers.empty()) {
+        throw std::runtime_error{"direct_failure_detector: trying to register a listener after endpoints were added"};
+    }
+
+    auto [it, _] = _listeners_liveness.try_emplace(threshold);
+    it->second.listeners.push_back(listener_info{id, shard});
+}
+
+void failure_detector::impl::remove_listener(listener_id id) {
+    // Linear search, but we don't expect a huge number of listeners and it's a rare operation.
+    for (auto ll_it = _listeners_liveness.begin(); ll_it != _listeners_liveness.end(); ++ll_it) {
+        auto& ll = ll_it->second;
+        for (auto it = ll.listeners.begin(); it != ll.listeners.end(); ++it) {
+            if (it->id == id) {
+                ll.listeners.erase(it);
+                if (ll.listeners.empty()) {
+                    // No more listeners with this threshold.
+                    // Remove the whole entry.
+                    _listeners_liveness.erase(ll_it);
+                }
+                return;
+            }
+        }
+    }
+}
+
+void failure_detector::add_endpoint(pinger::endpoint_id ep) {
+    if (_impl) {
+        _impl->send_update_endpoint(ep, endpoint_update::added);
+    }
+}
+
+void failure_detector::remove_endpoint(pinger::endpoint_id ep) {
+    if (_impl) {
+        _impl->send_update_endpoint(ep, endpoint_update::removed);
```

Comment:
Even if it would lead to shorter code, it would probably be more technically complex (a queue of functions is a weird thing to see, at least in C++). I think many would prefer this more verbose but easy to understand version.

Reply:
Queues of functions are part of a semaphore, for example, as a wait list. I was trying to decode the purpose of the related code, and it wasn't obvious at first that you're trying to serialize execution (is there any other reason for not executing things immediately in `remove_endpoint()`?). This intent is communicated more clearly by using a semaphore, that's the standard way used across the code base.

e.g.:

```
auto units = co_await _change_serializer.get_units();
remove_endpoint(ep);
```

On shutdown, you join by taking the units (semaphore is fair).

Then there's no need for the update fiber, condition variables, etc.

Reply:
> is there any other reason for not executing things immediately in remove_endpoint()?

Yes, I want the public API to be "synchronous" (note that `remove_endpoint` returns `void`, not `future<>`). So waiting on a semaphore also wouldn't work.

Reply:
Also, perhaps it will be useful to note that I used a queue here in v1 (it wasn't a queue of functions, it was a queue of <endpoint, bool> pairs) and during the review I think it was Avi who said we should avoid queues because they can form backlogs. The size of a map, on the other hand, is bounded by the number of different endpoints.

Reply:
To keep the API synchronous you can abandon the fiber and track it with a gate.

Given the use case, backlog growing is unlikely. Status changes should not be more frequent than processing of them. Is extra complexity worth avoiding it? I'm not convinced.

Reply:
> Given the use case, backlog growing is unlikely. Status changes should not be more frequent than processing of them.

Citing Avi from the mailing list review thread:
> They are rare, but when they happen, it can be in a flood (say due to stalls). Similarly reaction to changes can take long (same stalls). Hence, it's worth some extra effort to make it robust.

and his previous review:
>  Can't this form backlogs? There's no point in queuing false, true,
false, true. We only need to keep track of the last notified state and
current state (and work to make then the same). I recommend a
condition_variable instead of a queue.

I basically agreed and implemented the tracking of last state using the map.

Now since I already changed the used data structure once due to a reviewer's request, I want to avoid changing it again so another reviewer likes it more but the previous reviewer is displeased. For me personally it doesn't really matter, I mean I can code all day and try out 10 different data structures until everyone likes the code (probably impossible though), but I have to admit I feel it's a bit of a waste of time.

> To keep the API synchronous you can abandon the fiber and track it with a gate.

Then I replace serialization of requests with synchronicity, unless I don't see something (how would a gate guarantee serialization?)
I need both serialization (the set of endpoints corresponds to the order of `add`/`remove` calls) and synchronicity (no `future<>`). So I need a fiber. It can be written directly like a coroutine, as it is now (`update_endpoint_fiber`) or I could introduce a `future<>` state variable to which I attach incoming requests (like `_destroy_subscriptions`) - which would bring backlogs back in a hidden form and would be less readable.

> Is extra complexity worth avoiding it?

I feel like the reviewers of this pull request have been constantly overshooting their estimation of the complexity of this code, then proposing alternative solutions which to me do not look any bit less complex. Yes I'm biased because I wrote this. But I think there are many far more complex pieces of code in Scylla, and at least this one is well commented so you can actually understand the code by reading it.

Reply:
Ok, I rest my case for the sake of making progress.

>  Then I replace serialization of requests with synchronicity, unless I don't see something (how would a gate guarantee serialization?)

We use both a gate (for tracking background work) and a semaphore (for serialization of work). We have many such examples in the code, base, e.g.:

```
    return with_gate(_draining_eps_gate, [this, filter = std::move(filter)] () mutable {
        return with_semaphore(drain_lock(), 1, [this, filter = std::move(filter)] () mutable {

```

Reply:
> We use both a gate (for tracking background work) and a semaphore (for serialization of work). We have many such examples in the code, base, e.g.:

Ok, that would work (obviously with the small modification of discarding the `with_gate` future instead of waiting for it), although it would introduce "hidden" backlogs (waiters on the semaphore) and it would replace a map+condition_variable with a gate+semaphore, whether one is more complex than the other is debatable, I would say.

One thing I don't like in this pattern is that we depend on the fact that `with_gate` doesn't yield before invoking the body (if it did yield, then given two invocations of `with_gate`, their corresponding bodies could get reordered by the reactor). This is not specified in the docstring of `with_gate`, so it's basically an implementation detail and prone to changes (actually, maybe it's deducible from the type, since it's a template and the passed-in function doesn't necessarily have to return `future<>`?). And then of course one has to remember that semaphores give FIFO order.

> Ok, I rest my case for the sake of making progress.

Thanks.